src/env.rs

```diff
@@ -17,35 +24,46 @@ pub fn get_pam_env(handle: &mut PamHandle) -> Option<PamEnvList> {
 }
 
 impl PamEnvList {
-    pub fn to_vec(&mut self) -> Vec<(String, String)> {
-        let mut vec = Vec::new();
-
-        let mut idx = 0;
-        loop {
-            let env_ptr: *const *const c_char = unsafe { self.ptr.offset(idx) };
-            if unsafe { !(*env_ptr).is_null() } {
-                idx += 1;
-
-                let env = unsafe { CStr::from_ptr(*env_ptr) }.to_string_lossy();
-                let split: Vec<_> = env.splitn(2, '=').collect();
-
-                if split.len() == 2 {
-                    // Only add valid env vars (contain at least one '=')
-                    vec.push((split[0].into(), split[1].into()));
-                }
+    pub fn iter(&self) -> IterPamEnv {
+        IterPamEnv {
+            envs: self,
+            idx: 0,
+            ended: false,
+        }
+    }
+
+    pub fn as_ptr(&self) -> *const *const c_char {
+        self.ptr
+    }
+}
+
+impl<'a> Iterator for IterPamEnv<'a> {
+    type Item = &'a CStr;
+
+    fn next(&mut self) -> Option<&'a CStr> {
+        if self.ended {
+            return None;
+        }
+
+        unsafe {
+            let env_ptr: *const *const c_char = self.envs.ptr.offset(self.idx);
+            self.idx += 1;
+
+            if !(*env_ptr).is_null() {
```

Comment:
Do we need another check for `null` in the first redirection?

```suggestion
            if !env_ptr.is_null() && !(*env_ptr).is_null() {
```

Reply:
The original code was copied from your to_vec method. However the PAM documentation says this:

> The format of the memory is a malloc()'d array of char pointers, the last element of which is set to NULL. Each of the non-NULL entries in this array point to a NUL terminated and malloc()'d char string of the form: "name=value". 

The way I interpret that is the iterator should stop if the first indirection is null, and skip once if the second indirection is null.

Reply:
That's how I read it as well. Then the correct solution would be to just check `env_ptr.is_null()` right? We could leave in the second check just to be extra sure, but I would also be fine with omitting it.

Reply:
Either way It would not hurt to throw in a `debug_assert!()` check on the second indirection. I don't have time to test this today, but I'll push the change for now.

Reply:
That'll do I think. Thanks for your work!